cdap-security/src/main/java/io/cdap/cdap/security/auth/AccessTokenIdentityExtractor.java

```diff
@@ -0,0 +1,95 @@
+import java.util.Base64;
+
+/**
+ * The AccessTokenIdentityExtractor attempts to extract an {@link AccessToken} included in the Authorization header,
+ * validate it, then transform it into a {@link UserIdentityPair}.
+ */
+public class AccessTokenIdentityExtractor implements UserIdentityExtractor {
+  public static final String NAME = "AccessTokenIdentityExtractor";
+
+  private static final Logger LOG = LoggerFactory.getLogger(AccessTokenIdentityExtractor.class);
+
+  private final TokenValidator tokenValidator;
+  private final Codec<AccessToken> accessTokenCodec;
+  private final Codec<UserIdentity> userIdentityCodec;
+
+  @Inject
+  public AccessTokenIdentityExtractor(TokenValidator tokenValidator, Codec<AccessToken> accessTokenCodec,
+                                      Codec<UserIdentity> userIdentityCodec) {
+    this.tokenValidator = tokenValidator;
+    this.accessTokenCodec = accessTokenCodec;
+    this.userIdentityCodec = userIdentityCodec;
+  }
+
+  /**
+   * Extracts a {@link UserIdentityPair} from an HTTP request by validating its {@link AccessToken} in the Authorization
+   * header. The access token is an instance of {@link AccessToken} which is base64-serialized.
+   * @param request The HTTP Request to extract the user identity from
+   * @return the user identity backed by an access token
+   * @throws IdentityExtractionException on identity extraction failure
+   */
+  @Override
+  public UserIdentityPair extract(HttpRequest request) throws IdentityExtractionException {
+    String auth = request.headers().get(HttpHeaderNames.AUTHORIZATION);
+    String accessToken = null;
+    if (auth != null) {
+      int idx = auth.trim().indexOf(' ');
+      if (idx < 0) {
+        throw new IdentityExtractionException(IdentityExtractionException.Classification.ERROR_MISSING_CREDENTIAL,
+                                              "No access token found");
+      }
+
+      accessToken = auth.substring(idx + 1).trim();
+    }
+    if (accessToken == null || accessToken.length() == 0) {
+      throw new IdentityExtractionException(IdentityExtractionException.Classification.ERROR_MISSING_CREDENTIAL,
+                                            "No access token found");
+    }
+    TokenState state = tokenValidator.validate(accessToken);
+
+    if (!state.isValid()) {
+      throw new IdentityExtractionException(IdentityExtractionException.Classification.ERROR_INVALID_TOKEN,
+                                            String.format("Failed to validate access token with reason: %s", state));
+    }
+
+    byte[] decodedAccessToken = Base64.getDecoder().decode(accessToken);
+    try {
+      AccessToken accessTokenObj = accessTokenCodec.decode(decodedAccessToken);
+      UserIdentity userIdentityObj = accessTokenObj.getIdentifier();
+      byte[] encodedAccessTokenIdentifier = userIdentityCodec.encode(userIdentityObj);
+      return new UserIdentityPair(Base64.getEncoder().encodeToString(encodedAccessTokenIdentifier).trim(),
```

Comment:
Why need to call `trim`? I don't think Base64 encodeToString would have trailing space.

Reply:
Most of this I copied from the existing code, but I imagine it is to simplify handling any kinds of weird transmission issues where the header values have whitespace characters.